### PR TITLE
[TLS13] Add client auth

### PR DESCRIPTION
**Pull request by gpotter2**
- rebase of https://github.com/romain-perez/scapy/tree/add-tls13-client-auth
- PR finished
- unit tests added

### CODE REVIEW

**Comment by gpotter2** [comment]
Tests don't pass on travis because it uses `OpenSSL 1.0.2g` which doesn't support tls 1.3 (it needs `OpenSSL 1.1.1`).

Not sure how to fix that, but will look into it :/

**Comment by gpotter2** [comment]
@guedou @p-l- Ready for review

File: scapy/automaton.py [hunk 1]
```diff
@@ -698,7 +698,10 @@ class CommandMessage(AutomatonException):
     # Services
     def debug(self, lvl, msg):
         if self.debug_level >= lvl:
-            log_interactive.debug(msg)
+            if conf.interactive:
```
Comment (guedou) [inline scapy/automaton.py:701]: Is there a reason why we don't use logging for this?

File: scapy/layers/tls/handshake.py [hunk 2]
```diff
@@ -307,7 +311,6 @@ def tls_session_update(self, msg_str):
             for e in self.ext:
                 if isinstance(e, TLS_Ext_SupportedVersion_CH):
                     s.advertised_tls_version = e.versions[0]
-
```
Comment (guedou) [inline scapy/layers/tls/handshake.py:310]: Is this change needed?

File: scapy/layers/tls/handshake.py [hunk 3]
```diff
@@ -1056,6 +1059,7 @@ class TLS13CertificateRequest(_TLSHandshake):
                    _ExtensionsField("ext", None,
                                     length_from=lambda pkt: pkt.msglen -
                                     pkt.cert_req_ctxt_len - 3)]
+
```
Comment (guedou) [inline scapy/layers/tls/handshake.py:1062]: Is this change needed?

File: scapy/layers/tls/keyexchange.py [hunk 4]
```diff
@@ -154,7 +154,7 @@ class _TLSSignature(_GenericTLSSessionInheritance):
     #XXX 'sig_alg' should be set in __init__ depending on the context.
     """
     name = "TLS Digital Signature"
-    fields_desc = [SigAndHashAlgField("sig_alg", 0x0401, _tls_hash_sig),
+    fields_desc = [SigAndHashAlgField("sig_alg", 0x0804, _tls_hash_sig),
```
Comment (guedou) [inline scapy/layers/tls/keyexchange.py:157]: Is this change needed?

File: scapy/automaton.py (hunk 1 above)
Comment (gpotter2) [reply scapy/automaton.py:701]: I wasn't sure which logger to use. The interactive one won't work out of a shell

File: scapy/layers/tls/handshake.py (hunk 2 above)
Comment (gpotter2) [reply scapy/layers/tls/handshake.py:310]: This was just for consistency. It's written elsewhere without the space. Also it's weird

File: scapy/layers/tls/handshake.py (hunk 3 above)
Comment (gpotter2) [reply scapy/layers/tls/handshake.py:1062]: This one returned a flake8 error (i think)

File: scapy/automaton.py (hunk 1 above)
Comment (guedou) [reply scapy/automaton.py:701]: You need to configure the logger
```
import logging
logger = logging.getLogger("scapy")
logger.setLevel(logging.INFO)
logger.addHandler(logging.StreamHandler())
from scapy.all import * 
log_runtime.critical('error')
```

File: scapy/layers/tls/keyexchange.py (hunk 4 above)
Comment (guedou) [reply scapy/layers/tls/keyexchange.py:157]: ping

Comment (gpotter2) [reply scapy/layers/tls/keyexchange.py:157]: I think it's a safer default (that's what wikipedia says https://en.m.wikipedia.org/wiki/Probabilistic_signature_scheme ... romain did this).
previous was `sha256+rsa`, this is `sha256+rsaepss`.
I wonder if the "e" is a typo or not, since the 512 one has no e..

File: scapy/automaton.py (hunk 1 above)
Comment (gpotter2) [reply scapy/automaton.py:701]: In any case that's only used during the debug. It's super useful to debug the tls tests..
Where would we instantiace the logger otherwise.. In UTScapy?

Comment (guedou) [reply scapy/automaton.py:701]: Yes, that will be the better option.
